client/comms/wsconn.go

```diff
@@ -0,0 +1,349 @@
+package comms
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/dcrdex/server/comms/msgjson"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// pingWait is the maximum time in seconds to wait for a ping from
+	// the server. The client can be reconnected after this threshold since
+	// the websocket connection to the server is assumed to broken.
+	pingWait = 60 * time.Second
+
+	// writeWait is the maximum time in seconds to write to the connection.
+	writeWait = 5 * time.Second
+
+	// bufferSize is buffer size for the websocket connection's read channel.
+	readBuffSize = 128
+)
+
+// WsCfg is the configuration struct for initializing a WsConn.
+type WsCfg struct {
+	// The websocket host.
+	Host string
+	// The websocket api path.
+	Path string
+	// The maximum time in seconds to wait for a ping from the server.
+	PingWait time.Duration
+	// The maximum time in seconds to write to the connection.
+	WriteWait time.Duration
+	// The rpc certificate file path.
+	RpcCert string
+	// The rpc key file path.
+	RpcKey string
+	// Skip server cert verification, should only be used in testing.
+	InsecureSkipVerify bool
+	// ReconnectSync runs the needed reconnection synchronisation after
+	// a disconnect.
+	ReconnectSync func()
+	// The context cancel func.
+	Cancel context.CancelFunc
+}
+
+// WsConn represents a client websocket connection.
+type WsConn struct {
+	rID          uint64
+	cfg          *WsCfg
+	ws           *websocket.Conn
+	tlsCfg       *tls.Config
+	readCh       chan *msgjson.Message
+	sendCh       chan *msgjson.Message
+	req          map[uint64]*msgjson.Message
+	reqMtx       sync.RWMutex
+	writeMtx     sync.Mutex
+	connected    bool
+	connectedMtx sync.RWMutex
+	once         sync.Once
+	wg           sync.WaitGroup
+}
+
+// filesExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
+}
+
+// NewWsConn creates a client websocket connection.
+func NewWsConn(cfg *WsCfg) (*WsConn, error) {
+	if cfg.PingWait < 1 {
```

Comment:
1 needs to be time.Second, other wise this is comparing to 1 nanosecond I think.  https://play.golang.org/p/viWcYNfQ-q8